### ISSUE TITLE
GLES2: crash on wine 1.8.7 after GL ERROR - Shader Compiler (was glDebugMessageCallbackARB...)

### ISSUE THREAD

**Issue by efornara** [godotengine/godot#17284]
**Godot version:**

checkout: https://github.com/godotengine/godot/commit/6a6aab498eab5d35621079aa6b0548e74e00dcb6  + cherry-pick: https://github.com/efornara/godot/commit/82c7c8d745bda2b3d9426fdc171b477b98e409cd

**OS/device including version:**

- debian/stretch + wine-1.8.7
- Mesa DRI Intel(R) Sandybridge Mobile x86/MMX/SSE2

**Issue description:**

GLES2 crashes on wine.

**Steps to reproduce:**

If you enable GLES2 on windows by cherry picking my change, the following works:

    wine godot.windows.opt.tools.32.exe --video-driver GLES3

The following crashes:

    wine godot.windows.opt.tools.32.exe --video-driver GLES2

On my branch, I fixed it like this: https://github.com/efornara/godot/commit/dad7b14357e2d62dee2008490e949f5b62c1e8c6

You might not care about wine, but fixing it on wine had the by-product of fixing it on real hardware too (Windows 10, Core i5 2430M). See: https://www.reddit.com/r/godot/comments/81hcw7/unofficialexperimental_godot_301_build_android_es/

There is no point in making PRs (driver selection is different in my branch, and my fix is too conservative for master), but, when testing the proper fix, I reckon using wine to validate the solution might be useful.

**Comment by bruvzg** [comment]
`gl_context = memnew(ContextGL_Win(hWnd, opengl_3_context));` shouldn't work at all as there is not code for gl 2.1 context creation in [context_gl_win.cpp#L150](https://github.com/efornara/godot/blob/3.0-gles2/platform/windows/context_gl_win.cpp#L150).

`glDebugMessageControlARB` is loaded dynamically by GLAD and is NULL if not supported.

d34571254 with following patch works on macOS 10.13.3, Wine 3.0.
<details/>
  <summary>Patch</summary>

```patch
diff --git a/drivers/gles2/rasterizer_gles2.cpp b/drivers/gles2/rasterizer_gles2.cpp
index 187395d46..260bafcb8 100644
--- a/drivers/gles2/rasterizer_gles2.cpp
+++ b/drivers/gles2/rasterizer_gles2.cpp
@@ -148,8 +148,8 @@ void RasterizerGLES2::initialize() {
 	}
 
 // GLVersion seems to be used for both GL and GL ES, so we need different version checks for them
-#ifdef OPENGL_ENABLED // OpenGL 3.3 Core Profile required
-	if (GLVersion.major < 3) {
+#ifdef OPENGL_ENABLED // OpenGL 2.1 Profile required
+	if (GLVersion.major < 2) {
 #else // OpenGL ES 3.0
 	if (GLVersion.major < 2) {
 #endif
@@ -160,30 +160,26 @@ void RasterizerGLES2::initialize() {
 				"Fatal error: Insufficient OpenGL / GLES driver support");
 	}
 
-#ifdef __APPLE__
-// FIXME glDebugMessageCallbackARB does not seem to work on Mac OS X and opengl 2, this may be an issue with our opengl canvas..
-#else
 	if (true || OS::get_singleton()->is_stdout_verbose()) {
-		glEnable(_EXT_DEBUG_OUTPUT_SYNCHRONOUS_ARB);
-		glDebugMessageCallbackARB(_gl_debug_print, NULL);
-		glEnable(_EXT_DEBUG_OUTPUT);
+		if (glDebugMessageControlARB) {
+			glEnable(_EXT_DEBUG_OUTPUT_SYNCHRONOUS_ARB);
+			glDebugMessageCallbackARB(_gl_debug_print, NULL);
+			glEnable(_EXT_DEBUG_OUTPUT);
+		}
 	}
-#endif
 
 #endif // GLAD_ENABLED
 
-		// For debugging
+	// For debugging
 #ifdef GLES_OVER_GL
-#ifdef __APPLE__
-// FIXME glDebugMessageCallbackARB does not seem to work on Mac OS X and opengl 2, this may be an issue with our opengl canvas..
-#else
-	glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_ERROR_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
-	glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_DEPRECATED_BEHAVIOR_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
-	glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_UNDEFINED_BEHAVIOR_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
-	glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_PORTABILITY_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
-	glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_PERFORMANCE_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
-	glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_OTHER_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
-#endif
+	if (glDebugMessageControlARB) {
+		glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_ERROR_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
+		glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_DEPRECATED_BEHAVIOR_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
+		glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_UNDEFINED_BEHAVIOR_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
+		glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_PORTABILITY_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
+		glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_PERFORMANCE_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
+		glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_OTHER_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
+	}
 #endif
 	/* glDebugMessageInsertARB(
 			GL_DEBUG_SOURCE_API_ARB,
diff --git a/platform/windows/context_gl_win.cpp b/platform/windows/context_gl_win.cpp
index 8b57fdd9c..9e856f9b4 100644
--- a/platform/windows/context_gl_win.cpp
+++ b/platform/windows/context_gl_win.cpp
@@ -147,8 +147,17 @@ Error ContextGL_Win::initialize() {
 
 	wglMakeCurrent(hDC, hRC);
 
-	if (opengl_3_context) {
+	PFNWGLCREATECONTEXTATTRIBSARBPROC wglCreateContextAttribsARB = NULL; //pointer to the method
+	wglCreateContextAttribsARB = (PFNWGLCREATECONTEXTATTRIBSARBPROC)wglGetProcAddress("wglCreateContextAttribsARB");
+
+	if (wglCreateContextAttribsARB == NULL)
+	{
+		MessageBox(NULL, "Cannot get Proc Address for CreateContextAttribs", "ERROR", MB_OK | MB_ICONEXCLAMATION);
+		wglDeleteContext(hRC);
+		return ERR_CANT_CREATE;
+	}
 
+	if (opengl_3_context) {
 		int attribs[] = {
 			WGL_CONTEXT_MAJOR_VERSION_ARB, 3, //we want a 3.3 context
 			WGL_CONTEXT_MINOR_VERSION_ARB, 3,
@@ -157,31 +166,36 @@ Error ContextGL_Win::initialize() {
 			0
 		}; //zero indicates the end of the array
 
-		PFNWGLCREATECONTEXTATTRIBSARBPROC wglCreateContextAttribsARB = NULL; //pointer to the method
-		wglCreateContextAttribsARB = (PFNWGLCREATECONTEXTATTRIBSARBPROC)wglGetProcAddress("wglCreateContextAttribsARB");
-
-		if (wglCreateContextAttribsARB == NULL) //OpenGL 3.0 is not supported
-		{
-			MessageBox(NULL, "Cannot get Proc Address for CreateContextAttribs", "ERROR", MB_OK | MB_ICONEXCLAMATION);
+		HGLRC new_hRC = wglCreateContextAttribsARB(hDC, 0, attribs);
+		if (!new_hRC) {
 			wglDeleteContext(hRC);
-			return ERR_CANT_CREATE;
+			MessageBox(NULL, "Can't Create An OpenGL 3.3 Rendering Context.", "ERROR", MB_OK | MB_ICONEXCLAMATION);
+			return ERR_CANT_CREATE; // Return false
 		}
-
+		wglMakeCurrent(hDC, NULL);
+		wglDeleteContext(hRC);
+		hRC = new_hRC;
+	} else {
+		int attribs[] = {
+			WGL_CONTEXT_MAJOR_VERSION_ARB, 2, //we want a 2.1 context
+			WGL_CONTEXT_MINOR_VERSION_ARB, 1,
+			0
+		}; //zero indicates the end of the array
 		HGLRC new_hRC = wglCreateContextAttribsARB(hDC, 0, attribs);
 		if (!new_hRC) {
 			wglDeleteContext(hRC);
-			MessageBox(NULL, "Can't Create An OpenGL 3.3 Rendering Context.", "ERROR", MB_OK | MB_ICONEXCLAMATION);
+			MessageBox(NULL, "Can't Create An OpenGL 2.1 Rendering Context.", "ERROR", MB_OK | MB_ICONEXCLAMATION);
 			return ERR_CANT_CREATE; // Return false
 		}
 		wglMakeCurrent(hDC, NULL);
 		wglDeleteContext(hRC);
 		hRC = new_hRC;
+	}
 
-		if (!wglMakeCurrent(hDC, hRC)) // Try To Activate The Rendering Context
-		{
-			MessageBox(NULL, "Can't Activate The GL 3.3 Rendering Context.", "ERROR", MB_OK | MB_ICONEXCLAMATION);
-			return ERR_CANT_CREATE; // Return FALSE
-		}
+	if (!wglMakeCurrent(hDC, hRC)) // Try To Activate The Rendering Context
+	{
+		MessageBox(NULL, "Can't Activate The OpenGL Rendering Context.", "ERROR", MB_OK | MB_ICONEXCLAMATION);
+		return ERR_CANT_CREATE; // Return FALSE
 	}
 
 	wglSwapIntervalEXT = (PFNWGLSWAPINTERVALEXTPROC)wglGetProcAddress("wglSwapIntervalEXT");
diff --git a/platform/windows/os_windows.cpp b/platform/windows/os_windows.cpp
index 910010187..4243de377 100644
--- a/platform/windows/os_windows.cpp
+++ b/platform/windows/os_windows.cpp
@@ -30,6 +30,7 @@
 
 #include "os_windows.h"
 
+#include "drivers/gles2/rasterizer_gles2.h"
 #include "drivers/gles3/rasterizer_gles3.h"
 #include "drivers/windows/dir_access_windows.h"
 #include "drivers/windows/file_access_windows.h"
@@ -1069,12 +1070,19 @@ Error OS_Windows::initialize(const VideoMode &p_desired, int p_video_driver, int
 	}
 
 #if defined(OPENGL_ENABLED)
-	gl_context = memnew(ContextGL_Win(hWnd, true));
-	gl_context->initialize();
+	if (p_video_driver == VIDEO_DRIVER_GLES3) {
+		gl_context = memnew(ContextGL_Win(hWnd, true));
+		gl_context->initialize();
 
-	RasterizerGLES3::register_config();
+		RasterizerGLES3::register_config();
+		RasterizerGLES3::make_current();
+	} else {
+		gl_context = memnew(ContextGL_Win(hWnd, false));
+		gl_context->initialize();
 
-	RasterizerGLES3::make_current();
+		RasterizerGLES2::register_config();
+		RasterizerGLES2::make_current();
+	}
 
 	gl_context->set_use_vsync(video_mode.use_vsync);
 #endif

```

</details>

**Comment by Jayhost** [comment]
Godot is currently using compatibility mode. I think this is not good for Wine?

Anyways I'm using Core 3.3 for my local branch.
I commented out "glDebugMessageCallbackARB"
https://twitter.com/JadenHorst/status/969708724406415360

**Comment by bruvzg** [comment]
May be its good for Wine under Linux but not on macOS since macOS doesn't support compatibility mode at all.

**Comment by efornara** [comment]
@bruvzg Yes, that's what I proper patch would look like :-) What I can do is, once it lands on master, to test it on my version of wine to give us confidence.

I'm not sure why you say that `gl_context = memnew(ContextGL_Win(hWnd, opengl_3_context));` shouldn't work. I might not like it, and I might find your version cleaner, but this code runs fine with `opengl_3_context = false`:

https://github.com/godotengine/godot/blob/2.1.4-stable/platform/windows/context_gl_win.cpp

So, I reckon it's good enough for my branch too.

**Comment by bruvzg** [comment]
Yep I was wrong, it's creating temporary context first at [context_gl_win.cpp#L148](https://github.com/godotengine/godot/blob/2.1.4-stable/platform/windows/context_gl_win.cpp#L148), and should work if compatibility profile is supported.

Here's PR for extension support checks: #17285

**Comment by efornara** [comment]
Not sure if it makes sense to follow this one up, as it might be related to my version of wine. Feel free to close this issue.

Commit: https://github.com/godotengine/godot/commit/900384a622c7f3094540a43670421469e4ee0b06

GLES3 is fine as before. GLES2 crashes as before, and removing the call to glDebugMessageCallbackARB fixes the issue. However, I added a few print_line and it turns out that loading and calling the function was/is fine. The crashes actually happens after entering the function _gl_debug_print for the first time. The message was skipped in an early test. I commented out the checks, and now the program prints:

    OpenGL ES 2.0 Renderer: Mesa DRI Intel(R) Sandybridge Mobile x86/MMX/SSE2
    ERROR: _gl_debug_print: GL ERROR: Source: Shader Compiler       Type: Other    ID: 1    Severity:       Message: FS SIMD8 shader: 5 inst, 0 loops, 24 cycles, 0:0 spills:fills, Promoted 0 constants, compacted 80 to 48 bytes.
       At: drivers/gles2/rasterizer_gles2.cpp:114
    wine: Unhandled page fault on execute access to 0x182c9b1c at address 0x182c9b1c (thread 0009), starting debugger...

**Comment by bruvzg** [comment]
I can confirm this `_gl_debug_print` string conversion crashes on Wine 3.0 on Linux, `length` is 0 and `message` pointer is NULL, `length` check should be added. GLES3 is probably affected too.

**Comment by bruvzg** [comment]
Well `length` contains trash, actually it's known Wine bug with wrong calling convention: https://bugs.winehq.org/show_bug.cgi?id=38402

**Comment by bruvzg** [comment]
I do not think something like this should be merged into master, but this patch works for me on both Wine/Linux and Wine/macOS and keep debug messages enabled on real Windows.

<details>
 <summary>Patch</summary>

```diff
diff --git a/drivers/gles2/rasterizer_gles2.cpp b/drivers/gles2/rasterizer_gles2.cpp
index f4f42df11..755970f67 100644
--- a/drivers/gles2/rasterizer_gles2.cpp
+++ b/drivers/gles2/rasterizer_gles2.cpp
@@ -191,6 +191,15 @@ void RasterizerGLES2::initialize() {
 		}
 	}
 #endif
+
+#ifdef WINDOWS_ENABLED
+	HMODULE ntdll = GetModuleHandle("ntdll.dll");
+	if (ntdll && GetProcAddress(ntdll, "wine_get_version")) {
+		print_line("OpenGL debugging is broken on Wine, see https://bugs.winehq.org/show_bug.cgi?id=38402");
+		GLAD_GL_ARB_debug_output = false;
+	}
+#endif
+
 	if (true || OS::get_singleton()->is_stdout_verbose()) {
 		if (GLAD_GL_ARB_debug_output) {
 			glEnable(_EXT_DEBUG_OUTPUT_SYNCHRONOUS_ARB);
diff --git a/drivers/gles3/rasterizer_gles3.cpp b/drivers/gles3/rasterizer_gles3.cpp
index 3732fe77b..9919e62b8 100644
--- a/drivers/gles3/rasterizer_gles3.cpp
+++ b/drivers/gles3/rasterizer_gles3.cpp
@@ -160,6 +160,14 @@ void RasterizerGLES3::initialize() {
 				"Fatal error: Insufficient OpenGL / GLES driver support");
 	}
 
+#ifdef WINDOWS_ENABLED
+	HMODULE ntdll = GetModuleHandle("ntdll.dll");
+	if (ntdll && GetProcAddress(ntdll, "wine_get_version")) {
+		print_line("OpenGL debugging is broken on Wine, see https://bugs.winehq.org/show_bug.cgi?id=38402");
+		GLAD_GL_ARB_debug_output = false;
+	}
+#endif
+
 	if (OS::get_singleton()->is_stdout_verbose()) {
 		if (GLAD_GL_ARB_debug_output) {
 			glEnable(_EXT_DEBUG_OUTPUT_SYNCHRONOUS_ARB);
```

</details>

**Comment by efornara** [comment]
Just to be clear, I don't think my problem was in the formatting. The crash was there even when the function had an early exit.

In my branch I think I'll keep the debugging out, but if I end up enabling it, I'll probably settle for something like this:

    if (!length || !message)
        message = "?";

Since we are already discussing this, the enum to strings mappings don't look nice either. I'm sure they are all right according to the specs and they don't cause any problems (i.e. every possible value is properly handled), but I guess if the warning level is ever cranked up any sane compiler would flag them. I would also initialize them to "".

**Comment by bruvzg** [comment]
`length`/`message` check will not help, it's [Wine bug](https://bugs.winehq.org/show_bug.cgi?id=38402).
Here what happens in details:

Wine passes callback pointer directly to Linux OpenGL library, callback function is declared as `stdcall` (as it should be on Windows), but Linux OpenGL expect it to be `cdecl`.

- On 32-bit x86 parameters are passed in the same way in both conventions, function itself works correctly, but on callback function return both callee and caller cleans up the stack consequentially and app crashes due to stack corruption.

- On 64-bit x86 parameters are messed up completely:

| Parameter               | Callback reads from | OpenGL writes to |
| ----------------------- | ------------------- | ---------------- |
| GLenum source           | RCX                 | RDI              |
| GLenum type             | RDX                 | RSI              |
| GLuint id               | R8                  | RDX              |
| GLenum severity         | R9                  | RCX              |
| GLsizei length          | stack               | R8               |
| const GLchar *message   | stack               | R9               |
| const GLvoid *userParam | stack               | stack            |

This bug is [fixed](https://wine-staging.com/news/2016-02-23-release-1.9.4.html) in Wine "staging" branch.

**Comment by efornara** [comment]
Thank you @bruvzg for the explanation. I see now why it would crash even after an early return.

I agree it that patching godot to workaround a bug in wine would be a bit too much. I doubt anyone would normally run godot in wine.
